src/ioctls/kvm_vec.rs

```diff
@@ -0,0 +1,394 @@
+
+    /// Get the array entries
+    ///
+    fn entries(&self) -> &__IncompleteArrayField<F>;
+
+    /// Get the array entries as mut
+    ///
+    fn entries_mut(&mut self) -> &mut __IncompleteArrayField<F>;
+}
+
+/// An adapter that helps in treating a KvmArray more like an actual `Vec`.
+///
+pub struct KvmVec<T: Default + KvmArray<F>, F> {
+    // this variable holds the KvmArray structure. We use a `Vec<T>` To make the allocation
+    // large enough while still being aligned for `T`. Only the first element of `Vec<T>` will
+    // actually be used as a `T`. The remaining memory in the `Vec<T>` is for `entries`, which
+    // must be contiguous. Since the entries are of type `F` we must be careful to convert the
+    // desired capacity of the `KvmVec` from `F` to `T` when reserving or releasing memory.
+    kvm_array: Vec<T>,
+    // the capacity measured in elements of type `F`
+    capacity: usize,
+    elements_type: PhantomData<F>,
+}
+
+impl<T: Default + KvmArray<F>, F> KvmVec<T, F> {
+    /// Returns the capacity required by kvm_array in order to hold the provided number of  `F`
+    ///
+    fn num_elements_to_kvm_array_len(num_elements: usize) -> usize {
+        let size_in_bytes = size_of::<T>() + num_elements * size_of::<F>();
+        (size_in_bytes + size_of::<T>() - 1) / size_of::<T>()
+    }
+
+    /// Constructs a new KvmVec<T, F> that contains `num_elements` empty elements of type `F`
+    ///
+    pub fn new(num_elements: usize) -> KvmVec<T, F> {
+        let required_kvm_array_capacity =
+            KvmVec::<T, F>::num_elements_to_kvm_array_len(num_elements);
+
+        let mut kvm_array = Vec::with_capacity(required_kvm_array_capacity);
+        for _ in 0..required_kvm_array_capacity {
+            kvm_array.push(T::default())
+        }
+        *kvm_array[0].len_mut() = num_elements as u32;
+
+        KvmVec {
+            kvm_array,
+            capacity: num_elements,
+            elements_type: PhantomData,
+        }
+    }
+
+    /// Returns a reference to the actual KVM structure instance.
+    ///
+    pub fn as_original_struct(&self) -> &T {
+        &self.kvm_array[0]
+    }
+
+    /// Returns a mut reference to the actual KVM structure instance.
+    ///
+    pub fn as_mut_original_struct(&mut self) -> &mut T {
+        &mut self.kvm_array[0]
+    }
+
+    /// Get a pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_ptr(&self) -> *const T {
+        self.as_original_struct()
+    }
+
+    /// Get a mutable pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_original_struct()
+    }
+
+    /// Returns a mut `Vec<F>` that contains all the elements.
+    /// It is important to call `mem::forget` after using this vector. Otherwise rust will destroy it.
+    ///
+    unsafe fn as_vec(&mut self) -> Vec<F> {
+        let entries_ptr = self.as_mut_original_struct().entries_mut().as_mut_ptr();
```

Comment:
nit: any reason why you're always using the accessor methods instead of just working with the struct members directly?
```suggestion
        let entries_ptr = self.kvm_array[0].entries_mut().as_mut_ptr();
```

Reply:
I just find it more readable this way. `kvm_array[0]` looks weird. `as_mut_original_struct` gives more information on what we're actually accessing.

Reply:
ok